src/twisted/python/compat.py

```diff
@@ -355,7 +363,21 @@ def nativeString(s: AnyStr) -> str:
     return s
 
 
-def _matchingString(constantString, inputString):
+# WhateverString is different from typing.AnyString in that it can accept a
+# Union at typechecking time, but it will still be bound to a specific
+# stringlike class if that is possible at typecheck time. (AnyString,
+# on the other hand, *must* be determined by the typechecker to be str or
+# bytes, it can't be a Union[] at the call site.)
+#
+# We can't use a typevar with bound=Union[str, bytes] here, either,
+# since even if the caller passes in some subclass of str or bytes,
+# our return value will be a plain old str or bytes.
+_WhateverString = TypeVar("_WhateverString", str, bytes, Union[str, bytes])
+
+
+def _matchingString(
+    constantString: Union[bytes, str], inputString: _WhateverString
```

Comment:
I think this is cleaner with `@overload`

Reply:
I don't think `@overload` can provide the stated semantics? I could be wrong.

Reply:
```
@overload
def matchingString(constantString: bytes, inputString: str | ReadableBuffer) -> bytes: ...

@overload
def matchingString(constantString: str, inputString: str) -> str:

```
@overload
def matchingString(constantString: bytes, inputString: str | ReadableBuffer) -> bytes: ...

@overload
def matchingString(constantString: str, inputString: str | Readable Buffer) -> str: ...

def matchingString(constantString: str | bytes, inputString: str | ReadableBuffer) - > str | bytes:
    ...
```